README
NAME

    Dist::Zilla::Plugin::Git::Checkout - clone and checkout a Git
    repository

VERSION

    Version 0.004

SYNOPSIS

      # in dist.ini:
      [Git::Checkout]
      :version = 0.004
      repo = https://github.com/skirmess/dzil-inc.git

DESCRIPTION

    This plugin clones, or if it is already cloned, fetches and updates a
    Git repository.

    The plugin runs during the initialization phase, which is the same for
    bundles and plugins. You can check out a Git repository and load
    bundles or plugins from this repository.

      # in dist.ini
      [Git::Checkout]
      :version = 0.004
      repo = https://github.com/skirmess/dzil-inc.git

      ; add the lib directory inside the checked out Git repository to @INC
      [lib]
      lib = dzil-inc/lib

      ; this bundle is run from inside the checked out Git repositories lib
      ; directory
      [@BundleFromRepository]

    Git version 1.7.10 or later is required.

USAGE

 branch / revision / tag

    Available since version 0.004.

    Specifies what to check out. This can be a branch, a tag or a revision.
    Only one of these three options can be used.

    If none is specified it defaults to the branch returned by

        git symbolic-ref -q --short refs/remotes/origin/HEAD

    and if that doesn't exist to the branch master.

 dir

    The repositories workspace is checked out into this directory. This
    defaults to the basename of the repo without the .git suffix.

 push_url

    Allows you to specify a different push url for the repositories origin.
    One possible scenario would be if you would like to clone via https but
    push via ssh. This is optional.

 repo

    Specifies the address of the repository to clone. This is required.

SUPPORT

 Bugs / Feature Requests

    Please report any bugs or feature requests through the issue tracker at
    https://github.com/skirmess/Dist-Zilla-Plugin-Git-Checkout/issues. You
    will be notified automatically of any progress on your issue.

 Source Code

    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    https://github.com/skirmess/Dist-Zilla-Plugin-Git-Checkout

      git clone https://github.com/skirmess/Dist-Zilla-Plugin-Git-Checkout.git

AUTHOR

    Sven Kirmess <sven.kirmess@kzone.ch>

SEE ALSO

    Dist::Zilla, lib



Changes
Revision history for Dist-Zilla-Plugin-Git-Checkout

{{$NEXT}}

  - Switch license to the ISC license

0.004  2022-01-08 20:17:50 UTC

  INCOMPATIBLE CHANGES:
  - The checkout parameter is replaced with a branch, revision, and tag parameter

  Changes:
  - Replace Git::Wrapper with Git::Background

0.003  2021-07-16 20:58:22 UTC

  - After checking the brandch out, do a "git pull --ff-only" to actually update it

0.002  2020-08-10 16:33:27 UTC

  - Don't assume that root from Test::DZil is a Path::Tiny object
  - Set user.name and user.email in Git test repo beforing commiting to it from basic.t

0.001  2020-08-08 23:40:44 UTC

  - Initial release


t/lib/Local/RemoveOrigin.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2020-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::RemoveOrigin;

our $VERSION = '0.001';

use Moose;

with 'Dist::Zilla::Role::Plugin';

use Git::Background 0.003;
use Path::Tiny;

use namespace::autoclean;

our $NOK;

around plugin_from_config => sub {
    my $orig         = shift @_;
    my $plugin_class = shift @_;

    my $instance = $plugin_class->$orig(@_);

    $NOK = undef;

    {
        local $@;    ##no critic (Variables::RequireInitializationForLocalVars)

        my $ok = eval {
            my $workspace = path( $instance->zilla->root )->child('ws');
            Git::Background->run( qw(remote remove origin), { dir => $workspace } )->get;

            1;
        };

        if ( !$ok ) {
            $NOK = $@;
        }
    }

    return $instance;
};

__PACKAGE__->meta->make_immutable;

1;

__END__


t/lib/Local/Test/TempDir.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2020-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::Test::TempDir;

our $VERSION = '0.001';

use Carp                  qw(croak);
use Cwd                   qw(getcwd);
use File::Path 2.07       qw(remove_tree);
use File::Spec::Functions qw(catdir);

# Support Exporter < 5.57
require Exporter;
our @ISA       = qw(Exporter);    ## no critic (ClassHierarchies::ProhibitExplicitISA)
our @EXPORT_OK = qw(tempdir);

{
    my $temp_dir_base;

    sub _init {
        return if defined $temp_dir_base;

        my $root_dir;
        if ( exists $ENV{LOCAL_TEST_TEMPDIR_BASEDIR} ) {
            $root_dir = $ENV{LOCAL_TEST_TEMPDIR_BASEDIR};
            croak "env variables LOCAL_TEST_TEMPDIR_BASEDIR doesn't point to a valid directory: $root_dir" if !-d $root_dir;
        }
        else {
            $root_dir = getcwd();
            croak "Cannot get cwd: $!" if !defined $root_dir;
        }

        $temp_dir_base = catdir( $root_dir, 'tmp' );
        if ( !-d $temp_dir_base ) {
            mkdir $temp_dir_base or croak "Cannot create directory $temp_dir_base $!";
        }

        ( my $dirname = $0 ) =~ tr{:\\/.}{_};
        $temp_dir_base = catdir( $temp_dir_base, $dirname );
        if ( !-e $temp_dir_base ) {
            mkdir $temp_dir_base or croak "Cannot create directory $temp_dir_base $!";
        }
        elsif ( -l $temp_dir_base || !-d _ ) {
            croak "Not a directory $temp_dir_base";
        }
        else {
            remove_tree( $temp_dir_base, { safe => 0, keep_root => 1 } );
        }

        return;
    }

    my %counter;

    sub tempdir {
        my $label = defined( $_[0] ) ? $_[0] : 'default';
        $label =~ tr{a-zA-Z0-9_-}{_}cs;

        if ( exists $counter{$label} ) {
            $counter{$label}++;
        }
        else {
            $counter{$label} = '0';
        }

        $label = "${label}_$counter{$label}";

        _init();

        my $tempdir = catdir( $temp_dir_base, $label );
        mkdir $tempdir or croak "Cannot create directory: $!";

        return $tempdir;
    }
}

1;


t/lib/Local/UpdateRemote.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2020-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::UpdateRemote;

our $VERSION = '0.001';

use Moose;

with 'Dist::Zilla::Role::Plugin';

use Git::Background 0.003;
use MooseX::Types::Moose qw(Str);
use Path::Tiny;

use namespace::autoclean;

has repo => (
    is       => 'ro',
    isa      => Str,
    required => 1,
);

our $NOK;

around plugin_from_config => sub {
    my $orig         = shift @_;
    my $plugin_class = shift @_;

    my $instance = $plugin_class->$orig(@_);

    my $repo_path = $instance->repo;

    $NOK = undef;

    {
        local $@;    ##no critic (Variables::RequireInitializationForLocalVars)

        my $ok = eval {
            my $workspace = path($repo_path)->absolute->parent->child('update_ws');
            Git::Background->run( 'clone', $repo_path, $workspace->stringify )->get;
            my $git = Git::Background->new($workspace);

            my $file_c = path($workspace)->child('C');
            $file_c->spew('1087');

            $git->run( 'add', 'C' )->get;
            $git->run( 'commit', '-m', 'third commit' )->get;

            $git->run( 'tag', '-d', 'my-tag' )->get;
            $git->run( 'tag', 'my-tag' )->get;

            $git->run('push')->get;
            $git->run( 'push', '--tags', '-f' )->get;

            1;
        };

        if ( !$ok ) {
            $NOK = $@;
        }
    }

    return $instance;
};

__PACKAGE__->meta->make_immutable;

1;

__END__


corpus/create_test_repo.sh
#!/bin/bash

set -e

cd -- "$(dirname -- "$0")"
BASE_DIR=$PWD

rm -rf test test.bundle test2.bundle

mkdir test
cd test
git init .

git config user.email test@example.com
git config user.name Test

perl -MPath::Tiny -e 'path(q{A})->spew(q{5})'
git add A
git commit -m 'initial commit'
git tag v47

perl -MPath::Tiny -e 'path(q{A})->spew(q{7})'
git add A
git commit -m 'second commit'

git tag my-tag

git branch dev
git checkout dev

perl -MPath::Tiny -e 'path(q{A})->spew(q{11})'
perl -MPath::Tiny -e 'path(q{B})->spew(q{13})'
git add A B

git commit -m 'commit on dev branch'

git checkout master

git bundle create ../test.bundle --branches --tags HEAD


# ---
cd -- "$BASE_DIR"

rm -rf test

mkdir test
cd test
git init .

git config user.email test@example.com
git config user.name Test

perl -MPath::Tiny -e 'path(q{C})->spew(q{419})'
git add C
git commit -m 'initial commit'

git bundle create ../test2.bundle --branches --tags HEAD


Makefile.PL
use 5.006;
use strict;
use warnings;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
    ABSTRACT           => 'clone and checkout a Git repository',
    AUTHOR             => 'Sven Kirmess <sven.kirmess@kzone.ch>',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    DISTNAME         => 'Dist-Zilla-Plugin-Git-Checkout',
    LICENSE          => 'open_source',
    MIN_PERL_VERSION => '5.006',
    NAME             => 'Dist::Zilla::Plugin::Git::Checkout',
    PREREQ_PM        => {
        'Dist::Zilla::Role::BeforeRelease' => 0,
        'Git::Background'                  => '0.003',
        'Git::Version::Compare'            => 0,
        Moose                              => 0,
        'MooseX::Types::Moose'             => 0,
        'Path::Tiny'                       => 0,
        'Term::ANSIColor'                  => 0,
        'namespace::autoclean'             => 0,
        strict                             => 0,
        warnings                           => 0,
    },
    TEST_REQUIRES => {
        Carp                        => 0,
        Cwd                         => 0,
        'Dist::Zilla::Role::Plugin' => 0,
        Exporter                    => 0,
        'File::Path'                => '2.07',
        'File::Spec::Functions'     => 0,
        'Test::DZil'                => 0,
        'Test::Fatal'               => 0,
        'Test::MockModule'          => '0.14',
        'Test::More'                => '0.88',
        lib                         => 0,
    },
    VERSION => '0.003',
    test    => {
        TESTS => 't/*.t',
    },
);

my %FallbackPrereqs = (
    Carp                               => 0,
    Cwd                                => 0,
    'Dist::Zilla::Role::BeforeRelease' => 0,
    'Dist::Zilla::Role::Plugin'        => 0,
    Exporter                           => 0,
    'File::Path'                       => '2.07',
    'File::Spec::Functions'            => 0,
    'Git::Background'                  => '0.003',
    'Git::Version::Compare'            => 0,
    Moose                              => 0,
    'MooseX::Types::Moose'             => 0,
    'Path::Tiny'                       => 0,
    'Term::ANSIColor'                  => 0,
    'Test::DZil'                       => 0,
    'Test::Fatal'                      => 0,
    'Test::MockModule'                 => '0.14',
    'Test::More'                       => '0.88',
    lib                                => 0,
    'namespace::autoclean'             => 0,
    strict                             => 0,
    warnings                           => 0,
);

if ( !eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

if ( !eval { ExtUtils::MakeMaker->VERSION(6.52) } ) {
    delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
}

WriteMakefile(%WriteMakefileArgs);

# vim: ts=4 sts=4 sw=4 et: syntax=perl
